api/cluster-resources/aou-snippets-menu.js.template

```diff
@@ -1,38 +1,58 @@
+        customMenus = [rMenu];
+      }
+      snippets_menu.options['menus'] = snippets_menu.default_menus;
+      snippets_menu.options['menus'][0]['sub-menu'].splice(0, 0, ...customMenus);
 
-    // Show the menu which is relevant to the current kernel. Show both as a fallback.
-    let customMenus = [pyMenu, rMenu];
-    if (kernelName.startsWith("python")) {
-      customMenus = [pyMenu];
-    } else if (kernelName === "ir") {
-      customMenus = [rMenu];
-    }
-    snippets_menu.options['menus'] = snippets_menu.default_menus;
-    snippets_menu.options['menus'][0]['sub-menu'].splice(0, 0, ...customMenus);
+      // Include custom menu content from the nbconfigurator UI extension, if any.
+      // We primarily expect this to be used for manual testing of
+      // workbench-snippets code, so prefix it to differentiate from the menus
+      // installed by this extension.
+      const snippetsConfig = Jupyter.notebook.config.data.snippets || {};
+      if (snippetsConfig.include_custom_menu) {
+        const customExtMenu = JSON.parse(snippetsConfig.custom_menu_content);
+        customExtMenu['name'] = '[Custom] ' + customExtMenu['name'];
+        snippets_menu.options['menus'][0]['sub-menu'].splice(0, 0, customExtMenu);
+      }
+    };
 
-    // Include custom menu content from the nbconfigurator UI extension, if any.
-    // We primarily expect this to be used for manual testing of
-    // workbench-snippets code, so prefix it to differentiate from the menus
-    // installed by this extension.
-    const snippetsConfig = Jupyter.notebook.config.data.snippets || {};
-    if (snippetsConfig.include_custom_menu) {
-      const customExtMenu = JSON.parse(snippetsConfig.custom_menu_content);
-      customExtMenu['name'] = '[Custom] ' + customExtMenu['name'];
-      snippets_menu.options['menus'][0]['sub-menu'].splice(0, 0, customExtMenu);
-    }
+    // Session may be null at this point, e.g. for large notebooks, while the
+    // Jupyter UI is still doing initial rendering.
+    if (Jupyter.notebook.session && Jupyter.notebook.session.kernel_model) {
+      handleKernel(Jupyter.notebook.session.kernel_model.name);
+      console.log('loaded AoU snippets on initial load');
+    } else {
+      console.log('session data not yet available, listening for notebook load');
+      events.on('notebook_loaded.Notebook', () => {
+        // We can't avoid the initial Snippets menu insertion, so remove it here and render again.
+        $('.dropdown:contains("Snippets")').remove();
 
-    console.log('loaded custom AoU snippet-menus');
+        // Kernel name should now be available.
+        handleKernel(Jupyter.notebook.session.kernel_model.name);
+        snippets_menu.menu_setup(
+            snippets_menu.options.menus, snippets_menu.options.sibling, /* insertBeforeSibling */ false);
+        console.log('loaded AoU snippets on event');
+      });
+    }
+    console.log('loaded custom AoU snippet-menus extension');
```

Comment:
is there a version of the extension or other metadata you might want here?

Reply:
Not really. Best I could think of would be to plug in the git commit hash here, but I don't think it's worth the effort. There's only one source file here so it's easy enough to see which version you're looking at (and these logs go nowhere but the user's browser console).